daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+  if (!add_succeeded)
+    goto handle_failed_write;
+
+  g_object_unref (file_output_stream);
+  g_byte_array_unref (priv->write_buffer);
+  priv->write_buffer = g_byte_array_new ();
+
+  return TRUE;
+
+handle_failed_write:
+  g_object_unref (file_output_stream);
+  return FALSE;
+}
+
+/* Populates elems with a C array of elements that consume no more than the
+ * given number of bytes in total. Note that only the size of the underlying
+ * data is taken into consideration, not overhead. Elements are read in the same
+ * order in which they were stored; in other words, the circular file is FIFO.
+ * Only data that has been successfully saved with emer_circular_file_save will
+ * be read. Sets token to an opaque value that may be passed to
+ * emer_circular_file_remove to remove the elements that were read in a
+ * particular call to emer_circular_file_read. Tokens may not be reused, and any
+ * successful call to emer_circular_file_remove invalidates any outstanding
+ * tokens. If no elements were read but the read succeeded, then elems is set to
+ * NULL. Returns TRUE on success and FALSE on error.
+ */
+gboolean
+emer_circular_file_read (EmerCircularFile *self,
+                         GBytes         ***elems,
+                         gsize             data_bytes_to_read,
+                         gsize            *num_elems,
+                         guint64          *token,
+                         GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (priv->size == 0)
+    {
+      *elems = NULL;
+      *num_elems = 0;
+      *token = 0;
+      return TRUE;
+    }
+
+  GFileInputStream *file_input_stream =
+    g_file_read (priv->data_file, NULL /* GCancellable */, error);
+  if (file_input_stream == NULL)
+    return FALSE;
+
+  gboolean seek_succeeded =
+    g_seekable_seek (G_SEEKABLE (file_input_stream), priv->head, G_SEEK_SET,
+                     NULL /* GCancellable */, error);
+  if (!seek_succeeded)
+    {
+      g_object_unref (file_input_stream);
+      return FALSE;
+    }
+
+  GPtrArray *elem_array =
+    g_ptr_array_new_with_free_func ((GDestroyNotify) g_bytes_unref);
+
+  guint64 curr_data_bytes = 0;
+  guint64 curr_disk_bytes = 0;
+  GInputStream *input_stream = G_INPUT_STREAM (file_input_stream);
+  while (curr_disk_bytes < priv->size)
+    {
+      guint64 elem_size;
+      if (!read_elem_size (self, input_stream, &elem_size, error))
+        goto handle_failed_read;
+
+      guint64 next_data_bytes = curr_data_bytes + elem_size;
+      if (next_data_bytes > data_bytes_to_read)
+        break;
+
+      gboolean append_succeeded =
+        append_elem_to_array (self, input_stream, elem_array, elem_size, error);
+      if (!append_succeeded)
+        goto handle_failed_read;
+
+      curr_data_bytes = next_data_bytes;
+      curr_disk_bytes += sizeof (elem_size) + elem_size;
```

Comment:
Should add a comment here explaining that the `sizeof (elem_size)` is to include the length, which is only present in the data on disk, like commented on IRC.